sriov: Fix regression on waiting SRIOV

cathay4t opened the pull request:
The 20871bac introduced a regression causing SR-IOV verification error
when enabling large mount(60+) VF.

The root cause is we only retry verification
`VERIFY_RETRY_COUNT_SRIOV(60)` times when VF changed with missing ethernet
VF. Fixed by use `VERIFY_RETRY_COUNT_SRIOV` when we have VF count
changed.

In test with ixgbe, we noticed 60 seconds is not enough, hence increase
to 300 seconds for verification retry.

Integration test case included and manually tested on ixgbe card.

cathay4t committed:
sriov: Fix regression on waiting SRIOV

The 20871bac introduced a regression causing SR-IOV verification error
when enabling large mount(60+) VF.

The root cause is we only retry verification
`VERIFY_RETRY_COUNT_SRIOV(60)` times when VF changed with missing ethernet
VF. Fixed by use `VERIFY_RETRY_COUNT_SRIOV` when we have VF count
changed.

In test with ixgbe, we noticed 60 seconds is not enough, hence increase
to 300 seconds for verification retry.

Integration test case included and manually tested on ixgbe card.

Signed-off-by: Gris Ge <fge@redhat.com>

ffmancera reviewed (dismissed):
I trust that you tested this using SR-IOV NIC. Other than that, LGTM!

cathay4t commented:
I tried in ixgbe server multiple times, got random failure, let me fix it.

cathay4t committed:
apply: Perform pre-apply check before creating checkpoint

Do pre-apply check before creating the checkpoint when the desire does
not contains future VF(in this case the pre-check only happens after
SRIOV VF count changed).

It is hard to test this in auto test case. Manual checked using
overbook bond port.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t committed:
sriov: Handle unknown interface type for SRIOV

When changing SRIOV VF change without defining interface type, nmstate
will incorrectly treat it as non-SRIOV interface which lead to incorrect
verification retry count (5).

To support that use case, we need to use `MergedNetworkState` to resolve
the unknown interface type, then check whether we have SRIOV VF changes
to determine the timeout and verification retry count.

Unit test case included.
Integration test case `test_change_vf_from_62_to_63` does not have
interface type defined which is enough to reproduce this issue after
multiple retry.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t commented:
Two more patches has included to solve the use case when user desired VF changes without interface type defined.
